### PR TITLE
Remove "AOT" from LLVM target names.

### PR DESCRIPTION
**Pull request by ScottTodd**
TLDR: Starting a transition from legacy names like "dylib-llvm-aot" to just "llvm-cpu".

History: in the early days of the project, we used to have compilation paths for both LLVM JIT (Just-In-Time) and LLVM AOT (Ahead-Of-Time). As the project matured, we dropped the JIT target and refactored the AOT target a few times. On the runtime side, we went through a few iterations of how "executable loaders" were defined: "dylib" (system .so, .dylib, .dll), "static" (executable code linked into runtime code statically at build time), and "embedded" (platform agnostic ELF file).

The "dylib-llvm-aot" name was descriptive at the time, but it doesn't accurately represent what the project can do today. Embedded ELF is the default linking mode, and there is no JIT to contrast AOT against.

Notable renames in this change:
| Component | Old name | New Name | Notes |
| --- | --- | --- | --- |
| CMake | `IREE_TARGET_BACKEND_DYLIB_LLVM_AOT` | `IREE_TARGET_BACKEND_LLVM_CPU` | |
| CMake | `IREE_TARGET_BACKEND_WASM_LLVM_AOT` | `IREE_TARGET_BACKEND_LLVM_CPU_WASM` | Adds `WebAssembly` to `LLVM_TARGETS_TO_BUILD` |
| File names | `LLVMAOTTarget.h/.cpp` | `LLVMCPUTarget.h/.cpp` | |
| Compiler targets | (old names will remain for now) | `llvm-cpu` | |
| Flags | `dylib-llvm-aot` | `llvm-cpu` | Only where required to match the CMake option for now |
| Env vars | `IREE_LLVMAOT_SYSTEM_LINKER_PATH` | `IREE_LLVM_SYSTEM_LINKER_PATH` | |

In future changes:
* tests using `# REQUIRES: llvmaot` -> `llvmcpu`?
* target name usage `dylib-llvm-aot`, `llvm`, `cpu` -> `llvm-cpu`
* remove `dylib-llvm-aot` legacy name

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -193,9 +193,9 @@ option(IREE_TARGET_BACKEND_DEFAULTS "Sets the default value for all compiler tar
 cmake_dependent_option(IREE_TARGET_BACKEND_VMVX "Enables the 'vmvx' compiler target backend" ON ${IREE_BUILD_COMPILER} OFF)
 
 # Supported default target backends.
-cmake_dependent_option(IREE_TARGET_BACKEND_DYLIB_LLVM_AOT "Enables the 'dylib-llvm-aot' compiler target backend" ${IREE_TARGET_BACKEND_DEFAULTS} ${IREE_BUILD_COMPILER} OFF)
+cmake_dependent_option(IREE_TARGET_BACKEND_LLVM_CPU "Enables the 'llvm-cpu' compiler target backend" ${IREE_TARGET_BACKEND_DEFAULTS} ${IREE_BUILD_COMPILER} OFF)
```
Comment (ScottTodd) [inline CMakeLists.txt:196]:
```
CMake Error at build_tools/cmake/iree_check_test.cmake:245 (message):
  Unknown backend 'dylib-llvm-aot'.  Check IREE_TARGET_BACKEND_* options.

CMake Error at build_tools/cmake/iree_hal_cts_test_suite.cmake:62 (message):
  Unknown backend 'dylib-llvm-aot'.  Check IREE_TARGET_BACKEND_* options.
Call Stack (most recent call first):
  runtime/src/iree/hal/drivers/local_task/cts/CMakeLists.txt:15 (iree_hal_cts_test_suite)
```

Errrrr... I was hoping to make this change in a few steps, but these will need to be made together.

**Comment by benvanik** [comment]
omg yes! this is so much nicer!

`LLVMTarget.h/.cpp` should be LLVMCPUTarget (to differentiate from LLVMGPUTarget and be consistent)

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp [hunk 2]
```diff
@@ -35,7 +35,7 @@
 #include "mlir/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Export.h"
 
-#define DEBUG_TYPE "iree-llvmaot-target"
+#define DEBUG_TYPE "iree-llvm-target"
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:38]:
```suggestion
#define DEBUG_TYPE "iree-llvm-cpu-target"
```

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp [hunk 3]
```diff
@@ -137,9 +137,9 @@ static LogicalResult linkBuiltinLibrary(
   return success();
 }
 
-class LLVMAOTTargetBackend final : public TargetBackend {
+class LLVMTargetBackend final : public TargetBackend {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:140]:
```suggestion
class LLVMCPUTargetBackend final : public TargetBackend {
```

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp [hunk 4]
```diff
@@ -827,17 +827,23 @@ void registerLLVMAOTTargetBackends(
...
   };
 
+  // Preferred name.
+  // #hal.executable.target<"llvm-cpu", ...
+  static TargetBackendRegistration registration0("llvm-cpu", backendFactory);
+
+  // Abbreviated names.
   // #hal.device.target<"cpu", ...
-  static TargetBackendRegistration registration0("cpu", backendFactory);
+  static TargetBackendRegistration registration1("cpu", backendFactory);
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:839]: can we drop these? (or is this a followup?)

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp (hunk 4 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:839]: Follow-up, yeah. I'll go through and update our tests/samples/etc. that use the old names. I wanted to keep this diff smaller and give people some time to switch their scripts / downstream code over.

Should we keep the `cpu` and `llvm` names as shorthands, or only use `llvm-cpu`?

Comment (powderluv) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:839]: at least the `cpu` shorthand would be great.

**Comment by ScottTodd** [comment]
I just sent a PSA about this renaming: https://groups.google.com/g/iree-discuss/c/N-BBCURS7g4
